engine/consensus/matching/engine.go

```diff
@@ -880,6 +879,9 @@ func (e *Engine) sealResult(incorporatedResult *flow.IncorporatedResult) error {
 // have called checkSealing 1000 times without seeing the block (it's probably
 // no longer a valid extension of the state anyway).
 func (e *Engine) clearPools(sealedIDs []flow.Identifier) error {
+	if len(sealedIDs) == 0 {
+		return nil
```

Comment:
we are safe to exit earlier here

Reply:
⚠️ 
I am not sure we should exit here. While the node did not actively constructed new seals, other nodes might have and published them in blocks.More specifically:
* https://github.com/onflow/flow-go/blob/7ec477232d590bbd9da281e6fe903eb7d08db035/engine/consensus/matching/engine.go#L891 might have progressed compared to the last time this replica has executed `clearPools`

#### Context
Lets take a look at when mempool content expires. Conditions are different for the different mempools:
* `incorporatedResults` are fork specific. Therefore, as soon as a corresponding seal is generated, we can prune the `incorporatedResults` from the mempool 
* On the other hand, `approvals`, `seals`, `receipts`: the same entity can be included in blocks in different forks. Therefore, we can only remove an en entity when the corresponding seal has been finalized. Only then we are certain that no other can be extended, where we still might have to include the entity. 

#### Suggestion
If we want to streamline this logic:
* lets define 
   ```golang
   var sealedHeightAtLastCleaning uint64 \\ height of latest finalized sealed block, when mempool cleaning logic was last run
   ```
   Default value at boot would be `0`
* after retrieving the latest sealed and finalized block: https://github.com/onflow/flow-go/blob/7ec477232d590bbd9da281e6fe903eb7d08db035/engine/consensus/matching/engine.go#L891-L894 we can preemptively exit when:
   ```golang
	if len(sealedIDs) == 0 && sealed.Height == e.sealedHeightAtLastCleaning {
		return nil
	}
   ```